### PR TITLE
Implement the intensional difference members direct introduction rule

### PR DESCRIPTION
**Pull request by leungmanhin**

### CODE REVIEW

**Comment by leungmanhin** [comment]
@ngeiswei please have a look when you have time

File: opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  ;;
+  ;; min(attraction(A,X), 1 - attraction(B,X))
+  ;;
+  ;; where the attraction(A,X) is the strength of the TV of
+  ;;
+  ;; Attraction <TV>
+  ;;   A
+  ;;   X
+  (define (get-strength A-at B-at)
+    (min (cog-mean A-at) (- 1 (get-pattern-strength B-at))))
```
Comment (linas) [inline opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm]: This is OK, but I think you'll get vastly superior performance and more readable, easier-to-understand code if you use PredicateFormulas. See `/examples/atomspace/formulas.scm` for working examples.

Comment (ngeiswei) [reply opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm]: That's interesting @linas, I however don't understand how the TV of the conclusion is being set then, without invoking a grounded schema.

File: opencog/pln/rules/intensional/intensional-difference-member-introduction.scm [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  ;;
+  ;; min(attraction(A,X), 1 - attraction(B,X))
+  ;;
+  ;; where the attraction(A,X) is the strength of the TV of
+  ;;
+  ;; Attraction <TV>
+  ;;   A
+  ;;   X
+  (define (get-strength A-at B-at)
+    (min (cog-mean A-at) (- 1 (get-pattern-strength B-at))))
```
Comment (ngeiswei) [inline opencog/pln/rules/intensional/intensional-difference-member-introduction.scm]: I think I want to rethink that, I'll get back to you...

File: opencog/pln/rules/intensional/intensional-difference-member-introduction.scm [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  ;; Attraction <TV>
+  ;;   A
+  ;;   X
+  (define (get-confidence A-at B-at)
+    (min (cog-confidence A-at) (cog-confidence B-at)))
+
+  (let* ((MembLink conclusion)
+         (A (gar (car premises)))
+         (B (gdr (car premises)))
+         (memb (gar MembLink))
```
Comment (ngeiswei) [inline opencog/pln/rules/intensional/intensional-difference-member-introduction.scm]: @leungmanhin note that these 3 variables above are no longer used.

Comment (leungmanhin) [reply opencog/pln/rules/intensional/intensional-difference-member-introduction.scm]: Oh nice catch, just removed them :)

File: opencog/pln/rules/intensional/intensional-difference-member-introduction.scm (hunk 2 above)
Comment (ngeiswei) [reply opencog/pln/rules/intensional/intensional-difference-member-introduction.scm]: @leungmanhin I think it's better to replace `(get-pattern-strength B-at)` by `(cog-mean B-at)`, and thus you can get rid of the `get-pattern-strength` function.

I had introduced that to work around the lack of confidence support in the intensional-difference-direct-introduction rule (i.e. it doesn't take into account the confidence over the attraction links). But it's better then to improve that confidence support there, even if it ends being a hack too, than here.

Comment (leungmanhin) [reply opencog/pln/rules/intensional/intensional-difference-member-introduction.scm]: I see, sure just pushed the changes

File: opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm (hunk 1 above)
Comment (linas) [reply opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm]: Did you try running the example code?

Comment (ngeiswei) [reply opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm]: Yes, was able to set the TV on the `EvaluationLink` holding the `PredicateFormulaLink`, but what other atom types?

Comment (linas) [reply opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm]: > what other atom types?

I don't understand the question. Do you want to use GPN's (or DefinedPredNodes) with something other than evaluation links?

**Comment by ngeiswei** [comment]
Looks good @leungmanhin thanks.

File: opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm (hunk 1 above)
Comment (ngeiswei) [reply opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm]: For instance, given

```scheme
Inheritance A B
Inheritance B C
```

we want to calculate the TV of

```scheme
Inheritance A C
```

and set it on the inferred inheritance.

Comment (linas) [reply opencog/pln/rules/intensional/intensional-difference-members-direct-introduction.scm]: Oh, gack. I just looked more carefully; I misunderstood the original use-case.   You are abusing an "undocumented behavior" of the GSN to perform a side-effect, with the side-effect being to alter the TV of some other completely unrelated atom, having nothing at all to do with the GSN.  Yuck.  I guess you got lucky that this code works. Its certainly not the intended use case, and is fairly abusive of the system. 

I don't particularly see any clean way of doing this.  This is yet another fundamental architectural issue that we've ignored, and just hacked-around instead, and its clearly biting in the heinie now.  Ugh. They are starting to pile up, each one unfixed...

**Pull request merged**